src/lib/replication/service/raft_repl_service.cpp

```diff
@@ -100,22 +101,68 @@ void RaftReplService::start() {
     r_params.return_method_ = nuraft::raft_params::async_handler;
     m_msg_mgr->register_mgr_type(params.default_group_type_, r_params);
 
+    // Step 3: Load all the repl devs from the cached superblks. This step creates the ReplDev instances and adds to
+    // list. It is still not joined the Raft group yet
+    for (auto const& [buf, mblk] : m_sb_bufs) {
+        load_repl_dev(buf, voidptr_cast(mblk));
+    }
+    m_sb_bufs.clear();
+
+    // Step 4: Load all the raft group configs from the cached superblks. We have 2 superblks for each raft group
+    // a) repl_dev configuration (loaded in step 3). This block is updated on every append and persisted on next cp.
+    // b) raft group configuration (loaded in this step). This block is updated on every config change and persisted
+    // instantly
+    //
+    // We need to first load the repl_dev with its config and then attach the raft config to that repl dev.
+    for (auto const& [buf, mblk] : m_config_sb_bufs) {
+        raft_group_config_found(buf, voidptr_cast(mblk));
+    }
+    m_config_sb_bufs.clear();
+
+    // Step 5: Start the data and logstore service now. This step is essential before we can ask Raft to join groups etc
+    hs()->data_service().start();
+    hs()->logstore_service().start(hs()->is_first_time_boot());
+
+    // Step 6: Iterate all the repl dev and ask each one of the join the raft group.
```

Comment:
do we have any mechanism to check partial `raft_repl_dev`?
for example, when creating raft_repl_dev, the leader succeed adding the first follower to the group , but fail to add the second follower. 

now for this case , we just return an error to upper layer , not do any cleanup. i think it is the reponsibility of the raft repl severice to cleanup up all the partial raft group, even if restart, so that we can make sure no leak of raft_repl_dev

Reply:
We will be adding "remove_repl_dev" api as part of next work. There will be adding "destroy" the repl_dev functionality (which we can use for all error handling).